### ISSUE TITLE
HTTP api integration with tendermint

### ISSUE THREAD

**Issue by kansi** [bigchaindb/bigchaindb#1946]
Following from #1828 most of the APIs have been integrated. Below are some remaining apis which would need some thinking or they are no longer much useful because of tendermint.

<hr>

#### [Transaction status](https://docs.bigchaindb.com/projects/server/en/latest/http-client-server-api.html#get--api-v1-statuses?transaction_id=transaction_id) = **GET /api/v1/statuses?transaction_id={transaction_id}** `DEPRICATED`

Done?
- [x] pull request #1996 : add mode parameter to transaction endpoint, resolved #1951 and #1995

EDIT: The above PR allows the user to post a transaction in different modes. So, the user doesn't need to continously check the status of a transaction i.e. they can `POST` the transaction in `async` or `commit` mode.

The transaction status is crucial API for the client to figure out if and when a transaction has been successfully accepted by the network. The older api provided three `status` values for a posted transaction i.e. `backlog`, `undecided` and `valid`. To replicate the same we can take the following approach:
- create a `statuses` collection.
- when a txn in posted from bigcahindb to tendermint, add the txn to `statuses` collection with `status` as `backlog`
- when a txn passes `checkTx` the status is updated from `backlog` to `undecided`
- after `deliverTx` is executed successfully the txn in removed from the `statuses` collection as the txn is now present in the `transactions` collection which implies that its `valid`.

In order to check the status of a txn, the `transactions` collection is check first and if the txn in not found then it is searched in `statuses` collection else HTTP `404` status is returned.

**NOTE**: the `backlog` status of a txn is only available from the node to which the txn was sent by the client.

**NOTE**: an alternate way to achieve this is suggested here #1951 

<hr>

#### [Get Block](https://docs.bigchaindb.com/projects/server/en/latest/http-client-server-api.html#get--api-v1-blocks-block_id) = **GET /api/v1/blocks/{block_id}**

- [x] Done? See pull request #1970 

There are two ways to port this API
- Stored the block `id` a given transaction belongs to in the transaction itself. **NOTE** this would change the txn schema
- Store a list of txn `ids` in each document of `blocks` collection and then load the txns accordingly.

**NOTE**: Fields `"node_pubkey"`, `"voters"`, `"signature"` might not be avaiable because of integration with Tendermint. Although fields like `"node_pubkey"` and `"signature"` could be computed when abci calls `commit` to commit the block. 

<hr>

####  [Block votes](https://docs.bigchaindb.com/projects/server/en/latest/http-client-server-api.html#get--api-v1-votes?block_id=block_id) = **GET /api/v1/votes?block_id={block_id}** `DEPRICATED`

The block data provided by tendermint only contains the validators hash. This hash can verified by hashing the current known validator peers. Furthermore, the signature field would need to be dropped. It would be good to re-evaluate this api from the point of view of what tendermint is providing. 

<hr>

#### [Block status](https://docs.bigchaindb.com/projects/server/en/latest/http-client-server-api.html#get--api-v1-statuses?block_id=block_id) = **GET /api/v1/statuses?block_id={block_id}**  `DEPRICATED`

Since blocks are created only after `commit` it seems that this api is rendered unnecessary.

<hr>

#### [Block containing txn](https://docs.bigchaindb.com/projects/server/en/latest/http-client-server-api.html#get--api-v1-blocks?transaction_id=transaction_id&status=UNDECIDED|VALID|INVALID) 
= **GET /api/v1/blocks?transaction_id={transaction_id}&status={UNDECIDED|VALID|INVALID}**

- [x] Done? See pull request #2021 

`DEPRECATE`?

Since the blocks are created only after consensus it would imply that a given transaction can be found only in a `valid` single block.

If this endpoint is deprecated, will there be some other way to get the block ID (block height) of the block containing a particular transaction (with given transaction ID)?

**Comment by ttmc** [comment]
The transaction can't be changed by a BigchainDB node (e.g. to add the block id) because all transactions are signed and adding anything to the transaction would change the thing-to-be-signed, but no BigchainDB node has the necessary private keys to re-sign the transaction (nor should it).

If we create a `blocks` collection (i.e. a local one in each node), then we'd still be able to answer the query, "Which block contains the transaction with this transaction ID?" i.e. `GET /api/v1/blocks?transaction_id={transaction_id}`

I guess the `GET /votes` endpoint would also go away?

**Comment by kansi** [comment]
@ttmc I am also inclined on taking the second approach. Currently, we have a `blocks` collection wherein each document contains `height` and `app_hash` at that height. A new field `transaction_ids` could be added which would hold the list of transactions which were commited in the block.

>I guess the `GET /votes` endpoint would also go away?

Not sure what you are suggesting but don't we need any api to list the validators of a block?

**Comment by ttmc** [comment]
> ...don't we need any api to list the validators of a block?

If we can get that, then yes, that would be nice to have.

**Comment by ttmc** [comment]
For **Get Block**, the first suggestion is:

> Stored the block id a given transaction belongs to in the transaction itself. NOTE this would change the txn schema

As a general principle, nodes shouldn't change transactions in any way (even if they are able to do so). The second option seems better to me:

> Store a list of txn ids in each document of blocks collection and then load the txns accordingly.

**Comment by kansi** [comment]
Yes, the second solution is better as the first one violates the current transaction schema.

**Comment by ttmc** [comment]
Issues #2036 and #2037 will resolve all remaining TODOs on this issue, so I'm closing this issue now.
